lib/xboxrt/vcruntime/excpt.cpp

```diff
@@ -0,0 +1,201 @@
+        return DISPOSITION_CONTINUE_SEARCH;
+    }
+
+    *pDispatcherContext = pRegistrationFrame;
+    return DISPOSITION_COLLIDED_UNWIND;
+}
+
+extern "C" void _local_unwind2 (EXCEPTION_REGISTRATION_SEH3 *pRegistrationFrame, int stop)
+{
+    // Manually install exception handler frame
+    EXCEPTION_REGISTRATION nestedUnwindFrame;
+    nestedUnwindFrame.handler = reinterpret_cast<void *>(_nested_unwind_handler);
+    asm volatile (
+        "movl %%fs:0, %%eax;"
+        "movl %%eax, (%0);"
+        "movl %0, %%fs:0;"
+        : : "r"(&nestedUnwindFrame) : "eax", "memory");
+
+    const ScopeTableEntry *scopeTable = pRegistrationFrame->ScopeTable;
+
+    while (true) {
+        DWORD currentTrylevel = pRegistrationFrame->TryLevel;
+
+        if (currentTrylevel == TRYLEVEL_NONE) {
+            break;
+        }
+
+        if (stop != TRYLEVEL_NONE && currentTrylevel <= stop) {
+            break;
+        }
+
+        DWORD oldTrylevel = currentTrylevel;
+        pRegistrationFrame->TryLevel = scopeTable[currentTrylevel].EnclosingLevel;
+
+        if (!scopeTable[oldTrylevel].FilterFunction) {
+            // If no filter funclet is present, then it's a __finally statement
+            // instead of an __except statement
+            auto finallyFunclet = reinterpret_cast<void (*)()>(scopeTable[oldTrylevel].HandlerFunction);
+            finallyFunclet();
+        }
+    }
+
+    // Manually remove exception handler frame
+    asm volatile (
+        "movl (%0), %%eax;"
+        "movl %%eax, %%fs:0;"
+        : : "r"(&nestedUnwindFrame) : "eax", "memory");
+}
+
+extern "C" void _global_unwind2 (EXCEPTION_REGISTRATION_SEH3 *pRegistrationFrame)
+{
+    asm volatile (
+        "pushl $0;"
+        "pushl $0;"
+        "pushl 1f;"
+        "pushl %0;"
+        "call _RtlUnwind@16;"
+        "1:;" : : "r"(pRegistrationFrame));
+}
+
+extern "C" int _except_handler3 (_EXCEPTION_RECORD *pExceptionRecord, EXCEPTION_REGISTRATION_SEH3 *pRegistrationFrame, _CONTEXT *pContextRecord, EXCEPTION_REGISTRATION **pDispatcherContext)
+{
+    // Clear the direction flag - the function triggering the exception might
+    // have modified it, but it's expected to not be set
+    asm volatile ("cld");
+
+    if (pExceptionRecord->ExceptionFlags & (EXCEPTION_UNWINDING | EXCEPTION_EXIT_UNWIND)) {
+        // We're in an unwinding pass, so unwind all local scopes
+        // Unwinding must be done with the ebp of the frame in which we want to unwind
+        const DWORD scopeEbp = (DWORD)&pRegistrationFrame->_ebp;
+        asm volatile (
+            "pushl %%ebp;"
+            "movl %0, %%ebp;"
+            "pushl %2;"
+            "pushl %1;"
+            "call __local_unwind2;" // Call _local_unwind2(pRegistrationFrame, TRYLEVEL_NONE);
+            "addl $8, %%esp;"
+            "popl %%ebp;" : : "r"(scopeEbp), "r"(pRegistrationFrame), "i"(TRYLEVEL_NONE) : "eax", "ecx", "edx");
```

Comment:
Where does this clobber the named registers? Is the assembler not smart enough to check the clobbers of the called functions?

Should we reduce reliance on inline assembly here, by simply getting/setting ebp from inline assembly, but calling from the C side? Can we do this without the C side setting it back or being confused?

Reply:
> Where does this clobber the named registers? Is the assembler not smart enough to check the clobbers of the called functions?

Those are the caller-saved registers according to cdecl. I don't think the compiler can deduce calling conventions etc. from inline assembly calls - I definitely saw this pattern elsewhere when doing stuff like this.

> Should we reduce reliance on inline assembly here, by simply getting/setting ebp from inline assembly, but calling from the C side? Can we do this without the C side setting it back or being confused?

I'd rather avoid this - we can't guarantee that the compiler won't try to access some variable or something while `ebp` doesn't point to the function's frame.

Reply:
I had already suspected this was what was going on.
I'd rather see explicit saving of those registers around the call, or a comment as to why they are in the clobber list as part of the code. Otherwise someone might accidentally remove these in the future.